### PR TITLE
ZEPPELIN-2742. Unify Zeppelin's Config Storage

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?

This PR is for unifying the config storage of zeppelin. Currently we store it as local file, this PR introduce an interface `ConfigStorage` and one implementation based on hadoop filesystem api. User could add others in future.  

### What type of PR is it?
[ Improvement  | Refactoring]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-2742

### How should this be tested?
Unit test is added and manually tested

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

### CODE REVIEW

**Comment by felixcheung** [comment]
shouldn't this PR say "config storage" instead of "storage layer"?

**Comment by zjffdu** [comment]
make sense, PR and jira title changed

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java [hunk 1]
```diff
@@ -518,7 +522,17 @@ public boolean isNotebookPublic() {
   }
 
   public String getConfDir() {
-    return getRelativeDir(ConfVars.ZEPPELIN_CONF_DIR);
+    return getRelativeDir(getString(ConfVars.ZEPPELIN_CONF_DIR));
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java]: why this change? other places are calling `getRelativeDir(ConfVars.something)`

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java [hunk 2]
```diff
@@ -518,7 +522,17 @@ public boolean isNotebookPublic() {
...
+  }
+
+  public String getConfigFSDir() {
+    String fsConfigDir = getString(ConfVars.ZEPPELIN_CONFIG_FS_DIR);
+    if (StringUtils.isBlank(fsConfigDir)) {
+      LOG.warn(ConfVars.ZEPPELIN_CONFIG_FS_DIR.varName + " is not specified, fall back to local " +
+          "conf directory " + ConfVars.ZEPPELIN_CONF_DIR.varName);
+      return "file://" + getConfDir();
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java]: shouldn't this fallback to `getString(ConfVars.ZEPPELIN_HOME) + "/" + path`? This was the behavior before this change

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/user/AuthenticationInfo.java [hunk 3]
```diff
@@ -19,15 +19,14 @@
...
-
+import com.google.gson.Gson;
 import org.apache.commons.lang.StringUtils;
 import org.apache.zeppelin.common.JsonSerializable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.Gson;
+import java.util.ArrayList;
+import java.util.List;
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/user/AuthenticationInfo.java]: let's revert this file?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/FileSystemStorage.java [hunk 4]
```diff
@@ -149,7 +161,6 @@ public Void call() throws IOException {
 
   public synchronized <T> T callHdfsOperation(final HdfsOperation<T> func) throws IOException {
     if (isSecurityEnabled) {
-      UserGroupInformation.getLoginUser().reloginFromKeytab();
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/FileSystemStorage.java]: why this change?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java [hunk 5]
```diff
@@ -133,27 +115,14 @@ public void setRoles(String user, Set<String> roles) {
...
-    }
-
-    try {
-      File settingFile = new File(filePath);
-      if (!settingFile.exists()) {
-        settingFile.createNewFile();
+      try {
+        configStorage.save(info);
+      } catch (IOException e) {
+        LOG.error("Error saving notebook authorization file: " + e.getMessage());
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java]: LOG.error("Error saving notebook authorization file", e);

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/user/EncryptorTest.java [hunk 6]
```diff
@@ -17,10 +17,10 @@
 
 package org.apache.zeppelin.user;
 
-import java.io.IOException;
-
 import org.junit.Test;
 
+import java.io.IOException;
+
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/user/EncryptorTest.java]: hmm, lets revert?

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java (hunk 2 above)
Comment (zjffdu) [reply zeppelin-interpreter/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java]: Hadoop will treat it as hdfs file if no prefix with `file://`

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/FileSystemStorage.java (hunk 4 above)
Comment (zjffdu) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/FileSystemStorage.java]: It is not necessary to do relogin because hadoop rpc will do it.

**Comment by zjffdu** [comment]
Merge if no more comments

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java [hunk 7]
```diff
@@ -133,27 +115,14 @@ public void setRoles(String user, Set<String> roles) {
...
-    }
-
-    try {
-      File settingFile = new File(filePath);
-      if (!settingFile.exists()) {
-        settingFile.createNewFile();
+      try {
+        configStorage.save(info);
+      } catch (IOException e) {
+        LOG.error("Error saving notebook authorization file", e);
```
Comment (steveloughran) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java]: recommend adding `filePath` to that error string

Comment (zjffdu) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java]: The storage file path will be logged in `FileSystemConfigStorage`. And it doesn't expose api for getting the config file path. Actual `ConfigStorage` is a generic interface, it doesn't need to store into a file, the storage could be other storage like DB and etc.

Comment (steveloughran) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java]: OK. I'm just worrying about fielding the support calls where all you get is that stack trace.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/storage/FileSystemConfigStorage.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+  private FileSystemStorage fs;
+  private Path interpreterSettingPath;
+  private Path authorizationPath;
+  private Path credentialPath;
+
+  public FileSystemConfigStorage(ZeppelinConfiguration zConf) throws IOException {
+    super(zConf);
+    this.fs = FileSystemStorage.get(zConf);
+    this.fs.tryMkDir(new Path(zConf.getConfigFSDir()));
```
Comment (karth295) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/storage/FileSystemConfigStorage.java]: This is incorrect. `FileSystemStorage.get(zconf)` gives you an FS for zconf.getNotebookDir(). The notebook directory is *not* guaranteed to be on the same FS. Context: I tried to use the local `conf/` directory for zeppelin configuration, but put the notebook directory on a `gs://` path so that I could persistently save the notebook.

The same applies to interpreterSettingPath, authorizationPath, credentialPath.

FWIW, I think it's a mistake for FileSystemStorage/FileSystemNotebookRepo to use zconf.getNotebookDir(). Instead, there should be a separate config like zconf.getHcfsNotebookDir(). That way, each storage implementation has its own independent configuration. Right now, you cannot use FileSystemNotebookRepo with GitNotebookRepo, because they rely on the same configuration.

I'm happy to file JIRAs for these if you would like.
